### PR TITLE
ML classification notebooks

### PR DESCRIPTION
**Pull request by mcleonard**
For your review. Still a few things I need to fix.

### CODE REVIEW

File: notebooks/machine_learning/raw/tut8.ipynb [hunk 1]
```diff
@@ -0,0 +1,479 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Classification\n",
+    "\n",
+    "Up to this point you've been using data to make real-valued predictions such as home prices. This type of modeling is called **regression**, hence the \"Regressor\" part of `RandomForestRegressor`. In general, regression means you are trying to predict real numbers like prices, failure rates, energy consumption, etc.\n",
```
Comment (dansbecker) [inline notebooks/machine_learning/raw/tut8.ipynb]: This "real" stuff is more academic than we need.

Let's just say `So far you have predicted targets that are numbers.`

You've already said regression is predicting a number, so if you want to keep the last sentence, I'd rephrase it to `In general, regression means you are trying to predict real numbers like ...` to "Other examples of regression include ..."

Because we just said regression means predicting numbers.

File: notebooks/machine_learning/raw/tut8.ipynb [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+    "\n",
+    "Another common problem you'll see is making a choice between mutually exclusive outcomes. For example, spam detection is predicting if an email is \"spam\" or \"not spam\" based on the body text. Another common problem is identifying handwritten digits from images, choosing one of ten digits for any given image. This type of modeling is called **classification** because we are trying to choose one option from some set of classes.\n",
+    "\n",
+    "Classification is typically broken down further into binary and multiclass classification. Binary classification means there are two classes or conditions (\"spam\" or \"not spam\") while multiclass has more than two classes (ten digits). In general there are different approaches to the two types of classification, but most multiclass models will also work for binary problems.\n",
```
Comment (dansbecker) [inline notebooks/machine_learning/raw/tut8.ipynb]: You can shorten most of this paragraph to

```There are two types of classification: binary (choosing between two classes) and multiclass (choosing between more than two classes).```

File: notebooks/machine_learning/raw/tut8.ipynb [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+    "\n",
+    "It's straightforward to build classification models using what you already know about scikit-learn. Instead of `RandomForestRegressor` for regression, we can use `RandomForestClassifier` ([documentation here](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html#sklearn-ensemble-randomforestclassifier)). \n",
```
Comment (dansbecker) [inline notebooks/machine_learning/raw/tut8.ipynb]: We used to link to documentation frequently. 

When watching users, they frequently first questioned whether that meant they are supposed to go check it out, which they felt conflicted about.

And then they did go to it, and it sometimes sent them on a yak-shaving trip around the internet trying to figure out what some of it meant.  

With our goal being for users to get to basic competency quickly, we're avoiding link-outs to docs.

File: notebooks/machine_learning/raw/tut8.ipynb [hunk 4]
```diff
@@ -0,0 +1,479 @@
...
+    "\n",
+    "As an example of classification with `RandomForestClassifier`, I'll use [this dataset](https://www.kaggle.com/iabhishekofficial/mobile-price-classification#train.csv) for predicting price ranges for phones. The targets in the data have values:\n",
```
Comment (dansbecker) [inline notebooks/machine_learning/raw/tut8.ipynb]: I'd avoid linking out to the dataset too. If there's something they need to see, lets say it without making them branch off into a new browser tab. 

And if there isn't anything there they particularly need, we'll just keep with the flow of what we're saying.

File: notebooks/machine_learning/raw/tut8.ipynb [hunk 5]
```diff
@@ -0,0 +1,479 @@
...
+     ]
+    }
+   ],
+   "source": [
+    "# Predict classes given the validation features\n",
+    "pred_y = model.predict(val_X)\n",
+    "\n",
+    "# Calculate the accuracy as our performance metric\n",
+    "accuracy = metrics.accuracy_score(val_y, pred_y)\n",
+    "print(f\"Accuracy: {accuracy:.3f}\")"
```
Comment (dansbecker) [inline notebooks/machine_learning/raw/tut8.ipynb]: Most users have very little Python experience at this point, so they don't know f-strings.

We try to avoid showing things they don't understand or that make them worry they can't replicate what they've seen. Let's just print out full accuracy (despite the annoyingness of the extra sig-figures... so they don't feel like we're doing magic they can't replicate.

File: notebooks/machine_learning/raw/tut8.ipynb [hunk 6]
```diff
@@ -0,0 +1,479 @@
...
+    "print(f\"\\nNormalized confusion matrix:\")\n",
+    "for row in confusion:\n",
+    "    print(row / row.sum())"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<img src=\"https://i.imgur.com/idD0k8y.png\" alt=\"example confusion matrix\" width=400px>\n",
```
Comment (dansbecker) [inline notebooks/machine_learning/raw/tut8.ipynb]: I'd put a sentence like "In presentations, some people show this graphically like this:" above this graphic.

Following so closely on the output, I had to do a double take to see whether it was coming from the code... and a user may not know whether they are supposed to understand where this came from at this point.

File: notebooks/machine_learning/raw/tut8.ipynb [hunk 7]
```diff
@@ -0,0 +1,479 @@
...
+    "\n",
+    "The diagonal of the normalized confusion matrix gives us the **true positive rate** for each of the classes. This is the fraction of positive cases in the validation data the model predicted to be positive. This value is important because the accuracy metric can sometimes hide information you care about. Occasionally you'll have unbalanced classes where you have many more examples of one class compared to another. For example, it's typical that when you're trying to diagnose a disease with a machine learning model, most examples in your data won't have that disease. Your model could be very good at identifying when people don't have the disease (negative cases), but poor at identifying positive cases. \n",
```
Comment (dansbecker) [inline notebooks/machine_learning/raw/tut8.ipynb]: I'm skeptical of this characterization that things on the diagonal are the "True Positive Rate" in a multi-class setting.

As you say in the next sentence, the "true positive rate" is the fraction of actual positives that are predicted to be positive. But some parts of the diagonal are neither predicted nor actually "positive." 

In fact, in the multi-class case, the characterization of anything as a truly the "positive" class seems awkward.

Anyway, the diagonal could be called the "correctly classified" but not "True positive."

I think all of that is unnecessary though.

File: notebooks/machine_learning/raw/tut8.ipynb [hunk 8]
```diff
@@ -0,0 +1,479 @@
...
+    "\n",
+    "To illustrate this, assume negative cases make up 90% of your data and the classifier identifies negative cases at 95% accuracy.  Positive cases make up the other 10% of the data but your classifier only has a 25% accuracy for these cases. The overall accuracy would then be $90\\% * 95\\% + 25\\% * 5\\% = 88\\%$. The accuracy looks high and if that was the only metric you looked at you might think your classifier is doing pretty well. However, the goal is to identify when people have this disease, so 25% accuracy there is clearly not sufficient."
```
Comment (dansbecker) [inline notebooks/machine_learning/raw/tut8.ipynb]: I found it easy for my mind to wander in this paragraph.

What do you think about making a ThoughtQuestion out of this topic?

So the user gets to actively think instead of passively reading about it?

File: notebooks/machine_learning/raw/tut8.ipynb [hunk 9]
```diff
@@ -0,0 +1,479 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "You'll often see the true positive rate called the **recall** which you can calculate using `metrics.recall_score`. For binary classification you'll get one value while for multiclass classification, you'll get values for each of the classes. For more information about recall and associated metrics check out [the Wikipedia page](https://en.wikipedia.org/wiki/Precision_and_recall) and [the scikit-learn tutorial](https://scikit-learn.org/stable/auto_examples/model_selection/plot_precision_recall.html). "
```
Comment (dansbecker) [inline notebooks/machine_learning/raw/tut8.ipynb]: I think the likelihood a user remembers `metrics.recall_score` by the time they need this is extremely small. 

Which I think favors cutting the whole paragraph.

If all someone knows is "There are other measures of classification quality, like  `precision` and `recall`. Many data scientists don't remember their definitions, and you can look them up if you ever need them." I think that's enough for the user.

And that's much less cognitive load (and time) than reading docs or wikipedia.

File: notebooks/machine_learning/raw/tut8.ipynb [hunk 10]
```diff
@@ -0,0 +1,479 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Class probabilities \n",
+    "\n",
+    "The output from `model.predict` is a single class per input. However, classification models actually calculate a *probability distribution* over the classes. Using `model.predict` simply returns the class with the highest probability. This might not be ideal based on how the decision affects your metrics or downstream measures. To get the probabilities themselves, use the `.predict_proba` method."
```
Comment (dansbecker) [inline notebooks/machine_learning/raw/tut8.ipynb]: The first and third sentence are very similar. You can combine them into a single sentence.

File: notebooks/machine_learning/raw/ex8.ipynb [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Exercise: Classification\n",
+    "\n",
+    "Here you'll get some experience training a classification model yourself. What you'll do here is create a model that can determine if radio signals come from a pulsar. Pulsars are a rare type of neutron stars that produce radio signals we can detect on Earth. As the pulsars rotate, the beam of radio waves points directly at us, then moves away. This leads to a periodic signal that we can use to determine if the radio signal is actually from a pulsar or just noise.\n",
+    "\n",
+    "From the [dataset page on Kaggle](https://www.kaggle.com/pavanraj159/predicting-a-pulsar-star):\n",
```
Comment (dansbecker) [inline notebooks/machine_learning/raw/ex8.ipynb]: Most users just want to learn the ML. So I think it's plenty of detail alraedy in the intro paragraph. We don't need to say it again in the dataset page.

Broadly speaking, I worry a lot about the fact in UX design that users hate to read. So we should think of words as having some cost, and use them only if they move the user's knowledge forward (or they're necessary for the exercise to make sense).

File: notebooks/machine_learning/raw/ex8.ipynb [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+   "mimetype": "text/x-python",
+   "name": "python",
+   "nbconvert_exporter": "python",
+   "pygments_lexer": "ipython3",
+   "version": "3.6.8"
+  }
+ },
+ "nbformat": 4,
+ "nbformat_minor": 2
+}
```
Comment (dansbecker) [inline notebooks/machine_learning/raw/ex8.ipynb]: This exercise is pretty good. 

It'd be nice to add another ThoughtQuestion which is more abstract or conceptual.
We've heard a good response on those types of questions.  One example would be

"Think of a circumstance where you have an accuracy of 99%, and that seems surprisingly low"

You could then cut the imbalanced stuff from what the user passively reads... and give an answer to this thought question
```It would be surprising to have only 99% accuracy if a single outcome happens more than 99% of the time.
For example, if you are predicting whether a credit card transaction is fraudulent, and 99.9% of credit card transactions are legitimate, then you could get 99.9% accuracy by predicting everything is legitimate, even without a model.```

A user will remember this better if it's the response to a puzzle they've thought about, rather than it being stuck in the middle of a tutorial they are passively reading through."

**Comment by mcleonard** [comment]
@dansbecker I revised the notebooks, feel free to check it out when you have time.

**Comment by dansbecker** [comment]
Nice @mcleonard 

I made a couple small copy edits. I also merged a change from another branch to solve the build failure (which was caused by an unrelated change to a dataset the pandas course relied on).

I'll merge this to master. Then the next time kaggle updates the docker image, we can add these to the Learn page.

**Comment by dansbecker** [comment]
Damn. Just remembered we need to update the previous exercise so it doesn't say it's the end.  I can do that in a separate branch
